File: backend/routes/projects.js
const express = require("express");
const { isAuthenticated } = require("./auth");
const projects = require("../services/projects");

const router = express.Router();

router.get("/", isAuthenticated, (req, res) => {
  try {
    const allProjects = projects.getAllProjects();
    res.json(allProjects);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.get("/:id", isAuthenticated, (req, res) => {
  try {
    const project = projects.getProject(parseInt(req.params.id));
    if (!project) {
      return res.status(404).json({ error: "Project not found" });
    }
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post("/", isAuthenticated, (req, res) => {
  try {
    const { name, path, port, domain } = req.body;
    if (!name || !path || !port) {
      return res.status(400).json({ error: "name, path, and port required" });
    }
    const project = projects.addProject(name, path, port, domain);
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.put("/:id", isAuthenticated, (req, res) => {
  try {
    const project = projects.updateProject(parseInt(req.params.id), req.body);
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.delete("/:id", isAuthenticated, (req, res) => {
  try {
    const result = projects.deleteProject(parseInt(req.params.id));
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post("/:id/start", isAuthenticated, async (req, res) => {
  try {
    const result = await projects.startProject(parseInt(req.params.id));
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post("/:id/stop", isAuthenticated, (req, res) => {
  try {
    const result = projects.stopProject(parseInt(req.params.id));
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post("/:id/restart", isAuthenticated, async (req, res) => {
  try {
    const result = await projects.restartProject(parseInt(req.params.id));
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.get("/:id/logs", isAuthenticated, (req, res) => {
  try {
    const logs = projects.getProjectLogs(parseInt(req.params.id));
    res.json(logs);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: go-backend/internal/handlers/projects.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"github.com/kaysa/home-panel/internal/httpx"
	"github.com/kaysa/home-panel/internal/projects"
)

// Projects ports backend/routes/projects.js.
type Projects struct {
	Mgr *projects.Manager
}

func idParam(r *http.Request) int {
	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
	return id
}

func (p *Projects) List(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.GetAll())
}

func (p *Projects) Get(w http.ResponseWriter, r *http.Request) {
	proj, ok := p.Mgr.Get(idParam(r))
	if !ok {
		httpx.Error(w, http.StatusNotFound, "Project not found")
		return
	}
	httpx.JSON(w, http.StatusOK, proj)
}

func (p *Projects) Create(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name   string  `json:"name"`
		Path   string  `json:"path"`
		Port   float64 `json:"port"`
		Domain string  `json:"domain"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)
	if body.Name == "" || body.Path == "" || body.Port == 0 {
		httpx.Error(w, http.StatusBadRequest, "name, path, and port required")
		return
	}
	proj, err := p.Mgr.Add(body.Name, body.Path, int(body.Port), body.Domain)
	if err != nil {
		httpx.Error(w, http.StatusInternalServerError, err.Error())
		return
	}
	httpx.JSON(w, http.StatusOK, proj)
}

func (p *Projects) Update(w http.ResponseWriter, r *http.Request) {
	var body map[string]interface{}
	_ = json.NewDecoder(r.Body).Decode(&body)
	proj, _ := p.Mgr.Update(idParam(r), body)
	httpx.JSON(w, http.StatusOK, proj)
}

func (p *Projects) Delete(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.Delete(idParam(r)))
}

func (p *Projects) Start(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.Start(idParam(r)))
}

func (p *Projects) Stop(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.Stop(idParam(r)))
}

func (p *Projects) Restart(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.Restart(idParam(r)))
}

func (p *Projects) Logs(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, p.Mgr.Logs(idParam(r)))
}
